lib/checkstl.cpp

```diff
@@ -2610,10 +2610,25 @@ void CheckStl::useStlAlgorithm()
                 continue;
             const Token *bodyTok = tok->next()->link()->next();
             const Token *splitTok = tok->next()->astOperand2();
-            if (!Token::simpleMatch(splitTok, ":"))
-                continue;
-            const Token *loopVar = splitTok->previous();
-            if (!Token::Match(loopVar, "%var%"))
+            const Token* loopVar{};
+            if (Token::simpleMatch(splitTok, ":")) {
+                loopVar = splitTok->previous();
+                if (!Token::Match(loopVar, "%var%"))
+                    continue;
+            }
+            else if (Token::simpleMatch(splitTok, ";") && Token::simpleMatch(splitTok->astOperand2(), ";")) { // iterator-based loop
+                const Token* comp = splitTok->astOperand2()->astOperand1();
+                loopVar = Token::Match(comp, "%comp%") ? comp->astOperand1() : nullptr;
```

Comment:
The loopVar probably can't be just set to the iterator. I dont think we want to warn for cases like:

```cpp
auto g(const std::vector<int>& v) {
    std::vector<std::vector<int>::iterator> r;
    for (auto i = v.begin(); i != v.end(); ++i) {
        r.push_back(i);
    }
    return r;
}
```

It is possible to implement this with an algorithm but it requires special iterators(which C++ currently does not provide in the standard library). 

There could be other subtle differences where the iterator and value are not the same, and could be misinterpreted as the wrong algorithm. I think you need to search the loop for the dereference of this variable and set `loopVar` to that.

Reply:
But `loopvar` needs to have a varid, and `*i` or `i->` doesn't have one. The FP for your example should be avoided though.